### PR TITLE
Keeping sumPt from MET Significance calculation

### PR DESCRIPTION
**Pull request by danbarto**
#### PR description:

The sum of the unclustered energy is one of the inputs to MET significance, but is currently not stored in miniAOD, making recomputation of the significance and the MET covariance matrix difficult. With this PR, the calculated sum of the pTs of the unclustered particles is stored, and can be accessed with metSumPt(), similar to metSignificance().
This makes a later calculation, e.g. in nanoAOD as proposed in [this PR](https://github.com/cms-nanoAOD/cmssw/pull/443), redundant.

#### PR validation:

All tests have been run, and were passed.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoMET/METProducers/src/PFMETProducer.cc [hunk 1]
```diff
@@ -94,8 +94,9 @@ namespace cms {
     event.getByToken(rhoToken_, rho);
 
     //Compute the covariance matrix and fill it
+    double sumPt;
```
Comment (ahinzmann) [inline RecoMET/METProducers/src/PFMETProducer.cc]: Doesn't this need initialization = 0?

Comment (danbarto) [reply RecoMET/METProducers/src/PFMETProducer.cc]: You're right, thanks for the catch! It's initialized with 0 in the other producers, but somehow missing there. I'll fix it.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
Adding the JetMET POG conveners @ahinzmann @lathomas , since a specific MET reco contact isn't appointed yet

**Comment by ahinzmann** [comment]
Hi,
what is exactly the difference between met->sumEt and met->sumPt with this PR? (this is something non-experts will wonder).
If find the definition of sumEt here:
https://github.com/cms-sw/cmssw/blob/2ba5d421e10379d81760a899532b2c991b89c82c/RecoMET/METAlgorithms/src/METAlgo.cc#L44
So one difference is certainly that Et instead of Pt of particles is used (makes a difference for charged particles that are assigned a pi+- or lepton mass).
But is there any other difference in the computation of sumPt for signficance w.r.t. the general sumEt definition?
(depending on how significant the difference in definition is one may think of a more detailed name than sumPt)

**Comment by danbarto** [comment]
sumEt is the scalar sum of the Et of all PF candidates above the globalThreshold (which is 0 if I'm not wrong), while for sumPt overlap with leptons and jets is removed.

**Comment by ahinzmann** [comment]
so sumPt is rather sumPtUnclusered? Or is unclustered defined yet in another way?

**Comment by danbarto** [comment]
Yes, it's the measure of the unclustered energy in the MET significance. I agree that sumPtUnclustered is probably a less ambiguous name.

**Comment by slava77** [comment]
> Yes, it's the measure of the unclustered energy in the MET significance. I agree that sumPtUnclustered is probably a less ambiguous name.

https://github.com/cms-sw/cmssw/blob/81826f1ed7ccffddbd3d642838c057e31d309987/RecoMET/METAlgorithms/src/METSignificance.cc#L154-L157
this looks like more than unclustered

**Comment by danbarto** [comment]
Yes, jets with pt below the threshold (15 GeV by default) are also added to sumPt, and their associated PF candidates are removed to avoid double counting.

**Comment by ahinzmann** [comment]
So jets with pt<15 are treated as unclustered energy, so the name sumPtUnclustered could make sense.

**Comment by ahinzmann** [comment]
So my suggestion would be to rename to sumPtUnclustered.

**Comment by danbarto** [comment]
Fine with me, if there's no objection I'll rename it and update the PR.

**Comment by ahinzmann** [comment]
Just to confirm. Also @lathomas supports sumPtUnclustered.

**Comment by perrotta** [comment]
please test

File: DataFormats/PatCandidates/src/classes_def_objects.xml [hunk 2]
```diff
@@ -272,9 +272,11 @@
   <ioread sourceClass = "pat::Jet" version="[1-]" targetClass="pat::Jet" source="" target="daughtersTemp_">
     <![CDATA[daughtersTemp_.reset();]]>
   </ioread>
-  <class name="pat::MET"  ClassVersion="15">
-    <version ClassVersion="15" checksum="428901429"/>
+  <class name="pat::MET"  ClassVersion="17">
+    <version ClassVersion="17" checksum="2416242778"/>
```
Comment (perrotta) [inline DataFormats/PatCandidates/src/classes_def_objects.xml]: Please use consecutive numbers for ClassVersions, here "16".

File: DataFormats/PatCandidates/src/classes_def_objects.xml [hunk 3]
```diff
@@ -272,9 +272,11 @@
...
     <field name="corMap_" transient="true"/>
+   <version ClassVersion="16" checksum="2931227204"/>
```
Comment (perrotta) [inline DataFormats/PatCandidates/src/classes_def_objects.xml]: Intermediate versions can be neglected

File: DataFormats/PatCandidates/src/classes_def_objects.xml (hunk 2 above)
Comment (danbarto) [reply DataFormats/PatCandidates/src/classes_def_objects.xml]: done

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- "sumPtUnclustered" added to the MET DataFormat, according to the PR description and following review
- Jenkins tests pass and show no other differences than this additional double added to the dataformat

**Comment by santocch** [comment]
+1

**Comment by peruzzim** [comment]
A backport to 102X and 106X has to be discussed if you want this to be used in nanoAOD (and this implementation currently breaks the dataformat). Please follow up

**Comment by danbarto** [comment]
It would be desirable to have sumPtUnclustered in nanoAOD v7, so a backport will be needed.

**Comment by fabiocos** [comment]
@danbarto please provide it, so it can be merged once this feature is fully validated

**Comment by fabiocos** [comment]
+1

**Comment by danbarto** [comment]
@slava77 @fabiocos Do you have a suggestion/example on how to best keep sumPtUnclustered for later access in nanoAOD, without changing the data format (for the backport)? Thanks a lot for your help!

**Comment by slava77** [comment]
> @slava77 @fabiocos Do you have a suggestion/example on how to best keep sumPtUnclustered for later access in nanoAOD, without changing the data format (for the backport)? Thanks a lot for your help!

Addition of a new variable should be OK for a backport. The algorithm to fill and read from it will likely need to be enabled with an era modifier so that this reading is consistent. For now in 10_6_X you can use ```run2_miniAOD_devel``` to do that.

**Comment by peruzzim** [comment]
@slava77 I think that @danbarto was referring here to how to calculate it for using it in nanoAOD production without breaking MiniAOD consistency in ongoing campaigns, given that now the code is integrated with the producer that runs also during MiniAOD production.

**Comment by slava77** [comment]
> @slava77 I think that @danbarto was referring here to how to calculate it for using it in nanoAOD production without breaking MiniAOD consistency in ongoing campaigns, given that now the code is integrated with the producer that runs also during MiniAOD production.

shouldn't this be just to recompute the pat MET by rerunning the corresponding producer in the nanoAOD workflow?
The more convenient way would be to make a producer which can read the available pat::MET and fill just this variable via e.g. ModifiedObjectProducer infrastructure.

**Comment by peruzzim** [comment]
yes but backporting this will result in touching the MET dataformat...
sorry, maybe I'm not seeing the easy way out here... should one just backport the code and put the result somewhere transient?

**Comment by slava77** [comment]
> yes but backporting this will result in touching the MET dataformat...

as I mentioned above, adding a new variable should be OK as is. Consistent reading and filling of it in the production workflows will need use of appropriate modifiers.

**Comment by peruzzim** [comment]
This was not backported to 10_2_X, and is needed for NanoAODv7.
@danbarto would you have the chance to do it promptly?

#29180

**Comment by danbarto** [comment]
> This was not backported to 10_2_X, and is needed for NanoAODv7.
> @danbarto would you have the chance to do it promptly?
> 
> #29180

I should have time later today. Just to be sure I understand correctly, we need some era modifier to turn on/off filling of sumPtUnclustered for the backport?

**Comment by peruzzim** [comment]
@slava77 could you please provide some guidance on how to backport this to 10_2_X, technically? I'm also not sure. Thank you

**Comment by peruzzim** [comment]
After discussion with @silviodonato, we would propose to do the backport (to both 10_2_X and 10_6_X) as follows:

- change the MET dataformat as in the master branch
- add a configuration variable to `PATMETProducer`, let's call it `storeSumPtUnclustered`, that switches the filling of this variable by calling or not the `setMETSumPtUnclustered` function, and default it to `False`
- `run2_miniAOD_devel` modifier to set this variable `True`, for any possible future usage in re-MiniAOD campaigns
- also set it to `True` in the NanoAOD workflow (I can take care of this in the NanoAOD PR)

Would you agree?

**Comment by slava77** [comment]
> Would you agree?

please see https://github.com/cms-sw/cmssw/pull/28370#issuecomment-558644947
I think that a new variable can just be added without extra modifiers

**Comment by peruzzim** [comment]
ok, in this case we can just do a trivial backport without worrying of modifiers. @danbarto can you please prepare it?

**Comment by slava77** [comment]
> ok, in this case we can just do a trivial backport without worrying of modifiers. @danbarto can you please prepare it?

don't we also need a backport for 10_6_X?

**Comment by peruzzim** [comment]
yes indeed, not only for re-mini but also for backporting the nanoaodv7 configuration there (I'm working on that)

**Comment by danbarto** [comment]
Should I prepare the backport to 10_6_X?

**Comment by ahinzmann** [comment]
> Should I prepare the backport to 10_6_X?

Yes, please!

**Pull request merged**